tools/v8-json-to-junit.py

```diff
@@ -0,0 +1,121 @@
+import xml.etree.ElementTree as xml
+
+def IsExitCodeCrashing(exit_code):
+  if utils.IsWindows():
+    return 0x80000000 & exit_code and not (0x3FFFFF00 & exit_code)
+  return exit_code < 0 and exit_code != -signal.SIGABRT
+
+
+class JUnitTestOutput:
+  def __init__(self, test_suite_name):
+    self.root = xml.Element("testsuite")
+    self.root.attrib["name"] = test_suite_name
+
+  def HasRunTest(self, test_name, test_cmd, test_duration, test_failure):
+    test_case_element = xml.Element("testcase")
+    test_case_element.attrib["name"] = test_name
+    test_case_element.attrib["cmd"] = test_cmd
+    test_case_element.attrib["time"] = str(round(test_duration, 3))
+    if test_failure is not None:
+      failure_element = xml.Element("failure")
+      failure_element.text = test_failure
+      test_case_element.append(failure_element)
+    self.root.append(test_case_element)
+
+  def FinishAndWrite(self, f):
+    xml.ElementTree(self.root).write(f, "UTF-8")
+
+
+def Main():
+  test_results = json.load(sys.stdin)
+
+  # V8's JSON test runner only logs failing and flaky tests under "results". We
+  # assume the caller has put a large number for --slow-tests-cutoff, to ensure
+  # that all the tests appear under "slowest_tests".
+
+  failing_tests = {result["name"]: result for result in test_results["results"]}
+  all_tests = {result["name"]: result for result in test_results["slowest_tests"]}
+  passing_tests = {
+    name: result for name, result in all_tests.items() if name not in failing_tests
+  }
+
+  # These check that --slow-tests-cutoff was passed correctly.
+  assert len(failing_tests) + len(passing_tests) == len(all_tests)
+  assert len(all_tests) == len(test_results["slowest_tests"])
+
+  output = JUnitTestOutput("v8tests")
+
+  for name, failing_test in failing_tests.items():
+    failing_output = []
+
+    stdout = failing_test["stdout"].strip()
+    if len(stdout):
+      failing_output.append("stdout:")
+      failing_output.append(stdout)
+
+    stderr = failing_test["stderr"].strip()
+    if len(stderr):
+      failing_output.append("stderr:")
+      failing_output.append(stderr)
+
+    failing_output.append("Command: " + failing_test["command"])
+
+    exit_code = failing_test["exit_code"]
+    if failing_test["result"] == "TIMEOUT":
+      failing_output.append("--- TIMEOUT ---")
+    elif IsExitCodeCrashing(exit_code):
+      failing_output.append("exit code: " + str(exit_code))
+      failing_output.append("--- CRASHED ---")
+
+    output.HasRunTest(
+      test_name=name,
+      test_cmd=failing_test["command"],
+      test_duration=failing_test["duration"],
+      test_failure="\n".join(failing_output),
+    )
+
+  for name, passing_test in passing_tests.items():
+    output.HasRunTest(
+      test_name=name,
+      test_cmd=passing_test["command"],
+      test_duration=passing_test["duration"],
+      test_failure=None,
+    )
+
+  output.FinishAndWrite(sys.stdout.buffer)
+
+if __name__ == '__main__':
+  Main()
```

Comment:
For my understanding: this is a fixed-up version of the (now gone) tools/testrunner/local/junit_output.py script from upstream V8, right? Probably helpful to mention that in a top-level comment in this file.